### PR TITLE
Add new bitswap protocol from go-ipfs

### PR DESCRIPTION
**Pull request by rklaehn**
Also fix quite embarrassing bug in the proto serialization

### CODE REVIEW

File: src/bitswap/ledger.rs [hunk 1]
```diff
@@ -171,6 +171,9 @@ impl Into<Vec<u8>> for &Message<O> {
             payload.data = block.data().to_vec();
             proto.payload.push(payload);
         }
+        if !wantlist.entries.is_empty() {
+            proto.wantlist = Some(wantlist);
+        }
```
Comment (koivunej) [inline src/bitswap/ledger.rs]: Yeah this looks important, good find!

File: src/bitswap/bitswap_pb.proto [hunk 2]
```diff
@@ -3,29 +3,42 @@ syntax = "proto3";
...
+    }
+
     message Entry {
-      // the block cid (cidV0 in bitswap 1.0.0, cidV1 in bitswap 1.1.0)
-			bytes block = 1;
-      // the priority (normalized). default to 1
-			int32 priority = 2;
-      // whether this revokes an entry
-			bool cancel = 3;
+			bytes block = 1;		// the block cid (cidV0 in bitswap 1.0.0, cidV1 in bitswap 1.1.0)
```
Comment (dvc94ch) [inline src/bitswap/bitswap_pb.proto]: looks like mixed tabs and spaces?

**Comment by rklaehn** [comment]
I have just copied the proto file from the go-bitswap repo and removed some go specific extensions.

File: src/bitswap/ledger.rs (hunk 1 above)
Comment (dvc94ch) [reply src/bitswap/ledger.rs]: looks like it was introduced in #46

File: src/bitswap/bitswap_pb.proto (hunk 2 above)
Comment (rklaehn) [reply src/bitswap/bitswap_pb.proto]: yes, it is a basically verbatim copy of the file from go-ipfs. Did not bother to format it. Does it matter?

Comment (dvc94ch) [reply src/bitswap/bitswap_pb.proto]: I'd have formatted it but it's done now

**Pull request merged**